types/react/index.d.ts

```diff
@@ -883,7 +883,7 @@ declare namespace React {
     // based on the code in https://github.com/facebook/react/pull/13968
 
     // Unlike the class component setState, the updates are not allowed to be partial
-    type SetStateAction<S> = S | ((prevState: S) => S);
```

Comment:
I'm not sure I understand this change.  It looks like `S` was intended to capture a state type, so `SetStateAction` is either a state or a state transition.  After the change, it appears to say that if the state itself is a function (why would that be the case?) then it can't be a state and must be a state transition.  Am I reading that correctly?  If so, can you please explain a bit more about what the change was intended to accomplish?

Reply:
> it appears to say that if the state itself is a function (why would that be the case?) then it can't be a state and must be a state transition. Am I reading that correctly?

Yes, exactly.

https://reactjs.org/docs/hooks-reference.html#functional-updates

I don't think it's common case that S is a function, but if it is, S actually is `SetStateAction<ReturnType S>` not `SetStateAction<S>`. Even if it's always bad practice to make S a function, someone could accidentally do it and typescript won't complain about it anyway.

Reply:
Sorry, I'm not familiar with this library and I'm still not really following.  It looks like `SetStateAction` is only used by the overloads of `useState` and it looks like it's perfectly valid to pass a parameterless function to the first overload  - it will be used to compute the initial value and its return type will be passed to `SetStateAction`.  Perhaps you're worried about the case where someone pass a function that returns a function or explicitly specifies the state type argument as a function?  I agree that that would likely result in code that compiles but crashes at runtime, but I'm not sure how this change addresses that - it seems like it just prevents some forms of `setState` calls in those scenarios.

Reply:
```typescript
const [state, setState] = useState<number>(0);

setState(123); // ok, passing value, set to 123
setState(() => 456); // ok, passing functional update action, set to 456
```

`setState` here has type `(SetStateAction<S>) => void` and passing a _function returns a function_ as argument is perfectly valid when S is function.

The problem here is when `S' is a function. React **checks if an argument is a function** and uses the argument as a _functional update action_ or _value_ **depending on the check**.

```typescript
type MyState = (() => number) | null;
const [state, setState] = useState<MyState>(null);

setState(() => 123); // accidentally passing functional update action, set to 123 which is not S
setState(() => () => 456); // ok, passing functional update action, set to () => 456
```

So, when `S` is a function, it's not valid to pass a **_value_ directly** (of type `S` which is a function) on `setState` is not valid (since React assumes **it's a functional update action**).

This change prevents passing value directly when `S` is a function.

```typescript
type SetStateAction<S> =   S   |  ((prevState: S) => S);
//                       ^^^^^   ^^^^^^^^^^^^^^^^^^^^^^^^
//                       value   functional update action

////////////////////////////////////////////////////////////////////////////////////////

type SetStateAction<S> = (S extends (() => void) ? never : S) |  ((prevState: S) => S);
//                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^   ^^^^^^^^^^^^^^^^^^^^^^^^
//                       value not allowed if S is a function   functional update action
```

Reply:
> it's perfectly valid to pass a parameterless function to the first overload

I forgot, I just updated it now! https://github.com/DefinitelyTyped/DefinitelyTyped/pull/57563/commits/52baa2227dace36eb36d19b5af1af9a63ffda6e9

Reply:
Okay, I still don't completely understand why you'd want the state type to be a function type, but I'll just accept that.  If the goal is to detect function types, do you want to check something like `extends (...args: any[]) => any`?

Reply:
Good point, fixed at a834c213954f7e8ae15b7a17e1c4f9c95f110888. thanks :)

Reply:
>  do you want to check something like `extends (...args: any[]) => any`?

Sorry. I reverted [this change](https://github.com/DefinitelyTyped/DefinitelyTyped/commit/a834c213954f7e8ae15b7a17e1c4f9c95f110888) because it causes [type error](https://github.com/DefinitelyTyped/DefinitelyTyped/actions/runs/1561403313).

Reply:
I'm investigating the type error - it's not immediately clear why the change would cause that.

Reply:
https://github.com/microsoft/TypeScript/issues/47108

Reply:
So I guess stick with the current types for now.